File: problems/012-bounded-tar-stream/solutions/cpp/main.cpp
#include <cctype>
#include <iostream>
#include <string>
using namespace std;
using U = unsigned long long;
bool ok(const string &s) {
  if (s.empty() || s.front() == '/' || s.back() == '/')
    return false;
  size_t p = 0;
  while (p < s.size()) {
    size_t q = s.find('/', p);
    if (q == string::npos)
      q = s.size();
    string x = s.substr(p, q - p);
    if (x.empty() || x == "." || x == "..")
      return false;
    for (char c : x)
      if (!((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '.' ||
            c == '_' || c == '-'))
        return false;
    p = q + 1;
  }
  return true;
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n;
  U ln, lb;
  if (!(cin >> n >> ln >> lb))
    return 0;
  U off = 0, cnt = 0, used = 0;
  string pending;
  for (int i = 1; i <= n; i++) {
    U got, z, a, b;
    char t;
    string name, err;
    cin >> got >> t >> name >> z >> a >> b;
    if (got != off)
      err = "OFFSET";
    else if (a != b)
      err = "CHECKSUM";
    else if (string("FDGP").find(t) == string::npos)
      err = "TYPE";
    else if ((t == 'G' || t == 'P') && !pending.empty())
      err = "STATE";
    else if ((t == 'G' || t == 'P') && z != name.size() + 1)
      err = "META_SIZE";
    else if ((t == 'G' || t == 'P') && !ok(name))
      err = "PATH";
    else if ((t == 'F' || t == 'D') && !ok(pending.empty() ? name : pending))
      err = "PATH";
    else if (t == 'D' && z)
      err = "ENTRY_SIZE";
    else if (t == 'F' && (cnt == ln || z > lb - used))
      err = "LIMIT";
    if (!err.empty()) {
      cout << "REJECT " << i << ' ' << err << '\n';
      return 0;
    }
    off += 512 + ((z + 511) / 512) * 512;
    if (t == 'G' || t == 'P')
      pending = name;
    else {
      pending.clear();
      if (t == 'F') {
        cnt++;
        used += z;
      }
    }
  }
  if (!pending.empty())
    cout << "REJECT " << n + 1 << " STATE\n";
  else
    cout << "ACCEPT " << cnt << ' ' << used << ' ' << off << '\n';
}

File: problems/012-bounded-tar-stream/solutions/typescript/main.ts
import * as std from "std";
const input: string = std.in.readAsString();
let p: number = 0;
const nextToken = (): string => {
  while (p < input.length && input.charCodeAt(p) <= 32) p++;
  const start = p;
  while (p < input.length && input.charCodeAt(p) > 32) p++;
  return input.slice(start, p);
};
const n: number = Number(nextToken()),
  ln: bigint = BigInt(nextToken()),
  lb: bigint = BigInt(nextToken());
let off = 0n, cnt = 0n, used = 0n, pending: string | null = null, done = false;
const ok = (s: string): boolean =>
  !s.startsWith("/") && !s.endsWith("/") &&
  s.split("/").every((x) =>
    x !== "" && x !== "." && x !== ".." && /^[a-z0-9._-]+$/.test(x)
  );
for (let i = 1; i <= n; i++) {
  const got = BigInt(nextToken()),
    t = nextToken(),
    name = nextToken(),
    z = BigInt(nextToken()),
    x = BigInt(nextToken()),
    y = BigInt(nextToken());
  const meta = t === "G" || t === "P",
    actual = t === "F" || t === "D",
    eff = pending ?? name;
  let e: string | null = null;
  if (got !== off) e = "OFFSET";
  else if (x !== y) e = "CHECKSUM";
  else if (!"FDGP".includes(t)) e = "TYPE";
  else if (meta && pending !== null) e = "STATE";
  else if (meta && z !== BigInt(name.length + 1)) e = "META_SIZE";
  else if (meta && !ok(name)) e = "PATH";
  else if (actual && !ok(eff)) e = "PATH";
  else if (t === "D" && z !== 0n) e = "ENTRY_SIZE";
  else if (t === "F" && (cnt === ln || z > lb - used)) e = "LIMIT";
  if (e !== null) {
    std.out.puts(`REJECT ${i} ${e}\n`);
    done = true;
    break;
  }
  off += 512n + ((z + 511n) / 512n) * 512n;
  if (meta) pending = name;
  else {
    pending = null;
    if (t === "F") {
      cnt++;
      used += z;
    }
  }
}
if (!done) {
  std.out.puts(
    pending !== null
      ? `REJECT ${n + 1} STATE\n`
      : `ACCEPT ${cnt} ${used} ${off}\n`,
  );
}
